### PR TITLE
Fix #3693 Migration to React 16.10.1 and #4148: ESLint upgrade

### LINKED ISSUE #3693: Upgrade React to 16.x and also update related libraries
### Description
Following up investigation done for #3528 we want to upgrade React to the latest 16.x version and the following related libraries:

 - Babel (7.2.2)
 - Webpack (4.29.3)
 - react-redux (6.0.0)

We will also need to updgrade (to be compatible with React 16):
 - qrcode.react
 - react-color
 - react-nouislider
 - react-overlays
 - connected-react-router (used instead of react-router-redux)
 - react-selectize
 - react-widgets
 - reselect
 - react-sortable-items (created 16.x version on our fork)
 - react-pdf
 - ag-grid -> ag-grid-community + ag-grid-react
 - jsonix --> @boundlessgeo/jsonix

And also (to be compatible with webpack 4):
 - babel-loader
 - json-loader
 - copy-webpack-plugin
 - mini-css-extract-plugin (used instead of extract-text-webpack-plugin)
 - postcss, postcss-loader, postcss-prefix-selector
 - webpack-dev-server
 - webpack-parallel-uglify-plugin
 - webpack-cli (new)

Finally (for tests):
 - karma (4.3.0)
 - istanbul-instrumenter-loader
 - mocha (6.2.0 customized to fix an issue with uncaught exceptions)

### PR DESCRIPTION
**Pull request by mbarto**
## Description
This PR upgrades dependencies for a list of libraries, in particular:
 * ReactJS to 16.9.0 (latest)
 * ESLint to 6.3.0 (lates)

In addition, a lot of dependent libraries and build tools have been upgraded, the most important ones:
 * Redux
 * Babel
 * KarmaJS
 * Webpack

## Issues
 - #3693
 - #4148 

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)

 - [ ] Bugfix
 - [ ] Feature
 - [ ] Code style update (formatting, local variables)
 - [ ] Refactoring (no functional changes, no api changes)
 - [ ] Build related changes
 - [ ] CI related changes
 - [x] Other: libraries update

**Does this PR introduce a breaking change?** (check one with "x", remove the other)

 - [x] Yes, and I documented them in migration notes
 - [ ] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications:

Projects will need to update their dependencies. Some old ones could need some work to migrate.  In general most of the effort should be on build scripts and ESLint rules compliance.
Code using a mix of require / module.exports and import / export syntax IN THE SAME FILE will break.

**Other information**:

### CODE REVIEW

File: build/buildConfig.js [hunk 1]
```diff
@@ -35,6 +36,10 @@ module.exports = (bundles, themeEntries, paths, extractThemesPlugin, prod, publi
         'webpack-dev-server': 'webpack-dev-server/client?http://0.0.0.0:8081', // WebpackDevServer host and port
         'webpack': 'webpack/hot/only-dev-server' // "only" prevents reload on syntax errors
     }, bundles, themeEntries),
+    mode: "development",
```
Comment (MV88) [inline build/buildConfig.js]: Hi, great work, i think this will need to be overridden in prod-webpack.config

**Comment by MV88** [comment]
one other thing. i did not see changes to the migration guide. is it documented anywhere in the pr?

File: build/buildConfig.js (hunk 1 above)
Comment (mbarto) [reply build/buildConfig.js]: I changed buildConfig.js to switch using the prod flag

**Comment by mbarto** [comment]
> one other thing. i did not see changes to the migration guide. is it documented anywhere in the pr?

Still to be done

File: package.json [hunk 2]
```diff
@@ -67,16 +67,17 @@
     "redux-devtools": "3.4.0",
     "redux-devtools-dock-monitor": "1.1.2",
     "redux-devtools-log-monitor": "1.3.0",
-    "redux-immutable-state-invariant": "1.2.3",
+    "redux-immutable-state-invariant": "2.1.0",
```
Comment (MV88) [inline package.json]: i have found a bug when you run the app with debug=true
there is a problem here
```
var immutable = __webpack_require__(/*! redux-immutable-state-invariant */ "./node_modules/redux-immutable-state-invariant/dist/index.js")();
```
image

File: build/buildConfig.js [hunk 3]
```diff
@@ -142,31 +153,19 @@ module.exports = (bundles, themeEntries, paths, extractThemesPlugin, prod, publi
...
-                include: [
-                    paths.code,
-                    // the followings are here to skip issues caused by uglifyPlugin (see #4212)
-                    // TODO: a solution to this issue that uses only libs
-                    // like https://www.npmjs.com/package/uglify-js-es6#harmony when stable, terser, babel-minify ...
-                    path.join(paths.base, "node_modules", "query-string"),
-                    path.join(paths.base, "node_modules", "strict-uri-encode"),
-                    path.join(paths.base, "node_modules", "split-on-first")
-                ]
+                include: [paths.code, path.join(__dirname, "..", "node_modules", "query-string"), path.join(__dirname, "..", "node_modules", "strict-uri-encode"), path.join(__dirname, "..", "node_modules", "split-on-first")]
```
Comment (MV88) [inline build/buildConfig.js]: @mbarto Here you can use the fix made by @offtherailz  [here](https://github.com/geosolutions-it/MapStore2/pull/4268/files)

by replacing __dirname, "..",  with paths.base

Please would you mind to list them vertically? (big thanks)

Comment (mbarto) [reply build/buildConfig.js]: as you may know, the best way to make this happen is a PR

**Pull request merged**